### PR TITLE
[Inductor] [Quant] Enable QConv2d Unary int8-mixed-bf16 Lowering

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112486
* #112551
* __->__ #112550

**Summary**
- PR 5 for enabling Int8-Mixed-BF16 PT2E PTQ Quantization with Inductor https://github.com/pytorch/pytorch/issues/111640.
- Enable the QConv2d Unary int8-mixed-bf16 weight prepack and post grad lowering inside inductor.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/fx_passes/quantization.py [hunk 1]
```diff
@@ -48,12 +48,24 @@
     KeywordArg("w_dtype"),
 )
 
+dequantize_per_channel_to_bf16_weight_pattern = CallFunction(
+    prims.convert_element_type.default,
+    dequantize_per_channel_weight_pattern,
+    KeywordArg("autocast_weight_convert_dtype"),
+)
```
Comment (jgong5) [inline torch/_inductor/fx_passes/quantization.py:55]: maybe use a util function like `dtype_convert(pattern, dtype)` to simplify the places that need a type conversion on the output of the pattern?

Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py]: Thanks for the suggestion. Added `_generate_pattern_with_dtype_convert` as the util function to append pattern with dtype conversion.

**Review by jgong5 — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @eellison @jerryzh168, Could you kindly help to review these 3 PRs in this ghstack for the Inductor lowering of int8-mixed-bf16 quantization case?

**Comment by eellison** [comment]
@jerryzh168 want to review this one ?

File: torch/_inductor/fx_passes/quantization.py [hunk 2]
```diff
@@ -1162,16 +1264,26 @@ def _generate_qlinear_weight_prepack_patterns():
...
-        dequantize_per_tensor_activation_pattern, pass_number=0
-    )  # pass_number=0 to run before weight prepack
-
-    # Step 2: QConv weight prepack
-    weight_prepack_patterns = _generate_qconv_weight_prepack_patterns()
-    for weight_prepack_pattern in weight_prepack_patterns:
-        # Register to pass_number 1, so we can do dequant promotion in pass_number 0.
-        _register_qconv_weight_prepack_pass(weight_prepack_pattern, pass_number=1)
+    for dtype in [torch.float32, torch.bfloat16]:
+        # Step 1: Dequant promotion for int8-mixed-fp32/bf16
```
Comment (jerryzh168) [inline torch/_inductor/fx_passes/quantization.py:1268]: what does dequant promotion mean? can you write down the patterns or transformation?

File: torch/_inductor/fx_passes/quantization.py [hunk 3]
```diff
@@ -1162,16 +1264,26 @@ def _generate_qlinear_weight_prepack_patterns():
...
+        _register_dequant_promotion_pass(
+            dequantize_per_tensor_activation_pattern
+            if dtype == torch.float32
+            else _generate_pattern_with_dtype_convert(
+                dequantize_per_tensor_activation_pattern,
+                KeywordArg("autocast_act_dtype"),
```
Comment (jerryzh168) [inline torch/_inductor/fx_passes/quantization.py:1274]: can you just pass in dtype as an argument for `_generate_pattern_with_dtype_convert` and remove the if check?

File: torch/_inductor/fx_passes/quantization.py [hunk 4]
```diff
@@ -146,7 +167,12 @@ def generate_pattern_with_output_quant(computation_call):
                         aten.round.default,
                         CallFunction(
                             aten.mul.Tensor,
-                            computation_call,
+                            computation_call
+                            if dtype == torch.float32
+                            else _generate_pattern_with_dtype_convert(
+                                computation_call,
+                                KeywordArg("autocast_output_quant_dtype"),
+                            ),
```
Comment (jerryzh168) [inline torch/_inductor/fx_passes/quantization.py:175]: looks like this appears multiple times, refactoring a bit would be better I think

File: torch/_inductor/fx_passes/quantization.py [hunk 5]
```diff
@@ -18,6 +18,15 @@
 quantized_decomposed = torch.ops.quantized_decomposed
 quantized = torch.ops.quantized
 
+
```
Comment (jerryzh168) [inline torch/_inductor/fx_passes/quantization.py:21]: could you add some docs about how bfloat16 is handled (and its interaction with quantization)?

File: torch/_inductor/fx_passes/quantization.py [hunk 6]
```diff
@@ -756,30 +787,42 @@ def _register_quantization_lowerings():
...
-    if (
-        mul_node.target is aten.mul.Tensor
-        and sub_node.target is aten.sub.Tensor
-        and to_fp32_node.target is prims.convert_element_type.default
-        and len(list(mul_node.users)) > 1
-    ):
-        # dequant pattern has more than 1 users to be promoted
-        return True
-    return False
+def _is_valid_dequant_promotion_pattern(dtype=torch.float32):
```
Comment (jerryzh168) [inline torch/_inductor/fx_passes/quantization.py:790]: could you add some docs to explain dequant_promotion

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 7]
```diff
@@ -448,29 +448,38 @@ def __init__(
...
-            # 1. Dequant-Conv2D pattern matched in QConv2D weight prepack * 1
-            #    [convert_element_type_1, sub, mul_1, dequantize_per_channel, clone, convolution]
-            self.assertEqual(
-                counters["inductor"]["qconv2d_weight_prepack_matcher_count"], 1
-            )
-            self.assertEqual(
-                counters["inductor"]["qconv2d_weight_prepack_matcher_nodes"], 6
-            )
+        for int8_mixed_bf16 in (
+            [False, True] if torch.ops.mkldnn._is_mkldnn_bf16_supported() else [False]
```
Comment (jerryzh168) [inline test/inductor/test_mkldnn_pattern_matcher.py:460]: will it be better to split this to two tests?

File: test/inductor/test_mkldnn_pattern_matcher.py [hunk 8]
```diff
@@ -488,33 +497,35 @@ def __init__(
...
-                counters["inductor"]["qconv2d_weight_prepack_matcher_count"], 1
-            )
-            self.assertEqual(
-                counters["inductor"]["qconv2d_weight_prepack_matcher_nodes"], 6
-            )
-            # 2. QConv2D Unary fusion in post-grad fusion pass * 1
-            #    [qconv2d_pointwise_default, relu]
-            self.assertEqual(counters["inductor"]["qconv2d_unary_matcher_count"], 1)
-            self.assertEqual(counters["inductor"]["qconv2d_unary_matcher_nodes"], 2)
+        for int8_mixed_bf16 in (
```
Comment (jerryzh168) [inline test/inductor/test_mkldnn_pattern_matcher.py:510]: same here

**Review by jerryzh168 — APPROVED** [review]
LG overall, please add more some docs

File: torch/_inductor/fx_passes/quantization.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py:1268]: Thanks for the suggestion. Add the patterns and description for this pass.

File: torch/_inductor/fx_passes/quantization.py (hunk 3 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py:1274]: Thanks for the comments. The docs added and also the link to RFC for detail design. I will also update the tutorial for the usage of this feature.

File: torch/_inductor/fx_passes/quantization.py (hunk 5 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py:21]: Thanks for the comments. The docs added and also the link to RFC for detail design. I will also update the tutorial for the usage of this feature.

File: torch/_inductor/fx_passes/quantization.py (hunk 3 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py:1274]: Thanks and changed. It saves lots of redundant code.

File: torch/_inductor/fx_passes/quantization.py (hunk 4 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py:175]: Same as above and changed accordingly.

File: torch/_inductor/fx_passes/quantization.py (hunk 6 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/quantization.py:790]: Thanks for the suggestions. Docs added as in https://github.com/pytorch/pytorch/pull/112550#discussion_r1388459772

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 7 above)
Comment (leslie-fang-intel) [reply test/inductor/test_mkldnn_pattern_matcher.py:460]: Thanks for the suggestion and split.

File: test/inductor/test_mkldnn_pattern_matcher.py (hunk 8 above)
Comment (leslie-fang-intel) [reply test/inductor/test_mkldnn_pattern_matcher.py:510]: Thanks for the suggestion and split.

**Pull request closed without merging**